Modernize Docker workflow

felddy opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description

This PR updates the workflows to use more recent methods:
 - It uses the suite of Docker GitHub Actions (from CrazyMax).
 - It builds Docker images for more than one platform.
 - It handles making builds for branches, and tags them correctly.
 - It detects a release by the name of the git tag.
 - Nightly builds.

<!--- Describe your changes in detail -->

## 💭 Motivation and Context

I've been building this workflow based on some of the best-practices of the docker actions repos.

See here: https://github.com/docker/build-push-action#complete-workflow

Basically: If you change something in the repo, an image is going to pop out.  This is a good thing.  It makes testing branches a piece of cake. 

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## 🧪 Testing

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested for a few months in my personal project.
CI.

## 📷 Screenshots (if appropriate)

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

felddy committed:
Add script to generate cross-platform Dockerfile.

felddy committed:
Add git ignore for generated Dockerfile-x.

felddy committed:
Replace previous Docker workflows with a moderinzed, complete-workflow.

These changes are based on the "complete workflow" from
https://github.com/docker/build-push-action

Additionally it adds support for the new "workflow_dispatch" event type.

felddy committed:
Add CodeQL workflow. 

Needed to check tests written in Python.

felddy committed:
Update README publisher to use new Docker URL.

felddy committed:
Expand command option names to long form.

felddy committed:
Update Dockerfile to use opencontainers metadata and build-time labels.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -46,54 +66,148 @@ jobs:
-      - name: Cache testing environments
-        uses: actions/cache@v2
+      - name: Cache pip test requirements
+        uses: actions/cache@v1
```
```suggestion
        uses: actions/cache@v2
```

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -46,54 +66,148 @@ jobs:
-          key: "test-${{ runner.os }}-\
-            ${{ hashFiles('**/requirements-test.txt') }}-\
-            ${{ hashFiles('**/requirements.txt') }}"
+          key: "${{ runner.os }}-pip-test-\
+            ${{ hashFiles('**/requirements-test.txt') }}"
```
It's unclear to me why you used this old foo here:
```suggestion
          key: "test-${{ runner.os }}-\
            ${{ hashFiles('**/requirements-test.txt') }}-\
            ${{ hashFiles('**/requirements.txt') }}"
```

felddy replied:
Good catch.  I'm not sure how I regressed that.  I suspect I must have missed a PR in the descendant I was working with.  I'll try to be more careful about these changes in the future.   I don't want to stomp @mcdonnnj 's fine skeleton mechanizations.

jsf9k replied:
You are forgiven...this time.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -46,54 +66,148 @@ jobs:
+            ${{ hashFiles('**/requirements-test.txt') }}"
           restore-keys: |
-            test-${{ runner.os }}-
+            ${{ runner.os }}-pip-test-
+            ${{ runner.os }}-pip-
```
It's unclear to me why you used this old foo here:
```suggestion
            test-${{ runner.os }}-
```

felddy replied:
Oh, mi scusi!

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -46,54 +66,148 @@ jobs:
-      - name: Cache testing environments
-        uses: actions/cache@v2
+      - name: Cache pip test requirements
+        uses: actions/cache@v1
```
```suggestion
        uses: actions/cache@v2
```

felddy replied:
Tea for two!  🍵 🍵

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -46,54 +66,148 @@ jobs:
+          path: ${{ env.BUILDX_CACHE_DIR }}
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
```
We've been creating keys of the form `buildx-${{ runner.os }}` for a while now.  This is a change that @mcdonnnj encouraged some time ago.  I feel we should be consistent.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -102,4 +216,71 @@ jobs:
+        uses: actions/cache@v2
+        with:
+          path: ${{ env.BUILDX_CACHE_DIR }}
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
```
We've been creating keys of the form `buildx-${{ runner.os }}...` for a while now.  This is a change that @mcdonnnj encouraged some time ago.  I feel we should be consistent.

felddy replied:
Ah ha.  I found the reason I "reverted" to the ways of old.  It was good ol' cribbin' on my part from the Docker folks: 
https://github.com/docker/build-push-action#leverage-github-cache

jsf9k replied:
I assumed that was the case.  The Great and Powerful Oz knows all!

mcdonnnj replied:
I thought it might've been a copypasta incident. That's why we review though!

jsf9k commented on Dockerfile:
```diff
@@ -1,18 +1,11 @@
-LABEL maintainer="mark.feldhousen@trio.dhs.gov"
-LABEL vendor="Cyber and Infrastructure Security Agency"
-LABEL version=${VERSION}
+LABEL org.opencontainers.image.authors="mark.feldhousen@cisa.dhs.gov"
```
In case folks want to add more authors later, it might be good to have a comment here stating whether this string should be a comma-delimited list of email addresses or some other format.

felddy replied:
I've added comments to both places that generate labels for the container.
The [annotation](https://github.com/opencontainers/image-spec/blob/master/annotations.md) keys and values are defined in the [OCI Image Format Specification](https://github.com/opencontainers/image-spec#oci-image-format-specification).

That specific annotation, `org.opencontainers.image.authors`, is defined as `freeform string`: 
![tenor](https://user-images.githubusercontent.com/3229435/97900044-22757900-1d08-11eb-9874-b28dbde87aec.gif)

jsf9k reviewed (changes requested):
I think that these are great improvements, so strong work there @felddy!

That said, I think there are a few minor issues that should be cleared up before emerging.  Please see my inline comments.

In addition to that, and given that this is a skeleton repo that external folks may want to study before using, I think it _might_ make sense to add more comments in `.github/workflows/build.yml`.  The intent is less to detail what is being done by the code and more to guide the eye and help break up the giant workflow into digestible chunks (or "functions") that each perform a particular function.  That file is a little overwhelming at first glance, and it's difficult to find one's place again if one looks away at an email and comes back to it.

felddy committed:
Prevent image pushes when linter checks fail.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -220,7 +220,7 @@ jobs:
     name: "Build and push all platforms"
     runs-on: ubuntu-latest
-    needs: [prepare, test]
+    needs: [lint, prepare, test]
```
Nice catch!

jsf9k commented:
WRT commits 905799f, 3e7033b, and c506a21...I think GitHub gives you the option to combine multiple suggestions into a single commit.  I haven't had a chance to confirm that yet, but this would have been an excellent place to use that feature (assuming it exists).

Also, we should strive to use a better commit comment than the default offered by GitHub.  "Update .github/workflows/build.yml" is not very informative.

Not criticizing, just pointing out how we can better use our tools.

felddy committed:
Modify cache keys to follow team-established naming conventions.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k commented:
:notes: Workin' on my night builds :notes:
[![Night moves](https://img.youtube.com/vi/xH7cSSKnkL4/0.jpg)](https://www.youtube.com/watch?v=xH7cSSKnkL4&ab_channel=BobSegerVEVO)

felddy committed:
Add comments pointing to the opencontainers image-spec documentation.

felddy commented:
> WRT commits [905799f](https://github.com/cisagov/skeleton-docker/commit/905799f8e4f4cd2752ef67ead62f8de62470bc97), [3e7033b](https://github.com/cisagov/skeleton-docker/commit/3e7033b0c1134642f8ed9d52548cd73eb919740b), and [c506a21](https://github.com/cisagov/skeleton-docker/commit/c506a21fd19756f8962abbcedcd4c0491d9cccbc)...I think GitHub gives you the option to combine multiple suggestions into a single commit. I haven't had a chance to confirm that yet, but this would have been an excellent place to use that feature (assuming it exists).
> 
> Also, we should strive to use a better commit comment than the default offered by GitHub. "Update .github/workflows/build.yml" is not very informative.
> 
> Not criticizing, just pointing out how we can better use our tools.

I saw the "add to batch" button after I got through a couple of the suggestions.  I agree, those defaults suxors.  

Also see my comment in the description that I plan on squash-merging this PR since the commits are too exploratory (for my liking) due to the need to debug the workflow github-side.   I put that in there as I am open to preserving the individual commits if the team things there is value there.

mcdonnnj commented:
> 
> 
> > WRT commits [905799f](https://github.com/cisagov/skeleton-docker/commit/905799f8e4f4cd2752ef67ead62f8de62470bc97), [3e7033b](https://github.com/cisagov/skeleton-docker/commit/3e7033b0c1134642f8ed9d52548cd73eb919740b), and [c506a21](https://github.com/cisagov/skeleton-docker/commit/c506a21fd19756f8962abbcedcd4c0491d9cccbc)...I think GitHub gives you the option to combine multiple suggestions into a single commit. I haven't had a chance to confirm that yet, but this would have been an excellent place to use that feature (assuming it exists).
> > Also, we should strive to use a better commit comment than the default offered by GitHub. "Update .github/workflows/build.yml" is not very informative.
> > Not criticizing, just pointing out how we can better use our tools.
> 
> I saw the "add to batch" button after I got through a couple of the suggestions. I agree, those defaults suxors.
> 
> Also see my comment in the description that I plan on squash-merging this PR since the commits are too exploratory (for my liking) due to the need to debug the workflow github-side. I put that in there as I am open to preserving the individual commits if the team things there is value there.

Maybe do a rebase, mark the fluff as `fixup`, and force push to keep useful commits? That way we can avoid squashing the whole thing down, but also trim the fat so to speak.

jsf9k commented:
> Maybe do a rebase, mark the fluff as `fixup`, and force push to keep useful commits? That way we can avoid squashing the whole thing down, but also trim the fat so to speak.

100% agree with @mcdonnnj here.

felddy commented:
> In addition to that, and given that this is a skeleton repo that external folks may want to study before using, I think it _might_ make sense to add more comments in `.github/workflows/build.yml`. The intent is less to detail what is being done by the code and more to guide the eye and help break up the giant workflow into digestible chunks (or "functions") that each perform a particular function. That file is a little overwhelming at first glance, and it's difficult to find one's place again if one looks away at an email and comes back to it.

This is a good suggestion.  I'll add some detailed documentation about what is going on.   Mainly for future-me.  Because I'm sure this will run for six months then fail, and I'll come back all Gandalf-in-Moria when it breaks.

felddy commented:
> Maybe do a rebase, mark the fluff as `fixup`, and force push to keep useful commits? That way we can avoid squashing the whole thing down, but also trim the fat so to speak.

You like it when GitHub goes down?!   Cause this is how GitHub goes down.   😉 
I'll attempt this!  Keep an eye on: https://www.githubstatus.com/

jsf9k commented:
> You like it when GitHub goes down?! Cause this is how GitHub goes down. 
> I'll attempt this! Keep an eye on: https://www.githubstatus.com/

I do this sort of thang tous les temps.  C'est une chose quotidienne pour moi.  You and GitHub will be OK! :fr:

felddy commented:
Interactive rebase == Great Success!!   

> Jesus H. Christ we're still here! 
https://www.youtube.com/watch?v=ir5bTic17k4

felddy committed:
Add comments to help document and break up a monolitic file.

jsf9k reviewed (approved):
This PR looks great.  Thanks for resolving all the thangs.

dav3r commented on .github/workflows/build.yml:
```diff
@@ -46,40 +69,179 @@ jobs:
+    #     Builds targeting the default branch will be tagged with `:edge`.
+    #
+    #     Builds from other branches will be tagged with the branch name.
+    #     Solidi `/` in branch names are replaced with hyphens `-` in the Docker
```
`Solidi`?  What do [ancient Roman gold coins](https://www.merriam-webster.com/dictionary/solidi) have to do with this?

felddy replied:
`Solidus` is the proper name for [that Unicode character](https://unicodelookup.com/#solidus/1).   
`Hyphen` should technically be [`hyphen-minus`](https://unicodelookup.com/#hyphen-minus/1).    
Good catch!

dav3r replied:
TIL about the `solidus` Unicode character! 👍

dav3r reviewed (approved):
Super sweet improvements- can't wait to get this sauce cooking! ⚙️ 

I discovered a lone typo in a comment, but everything else looks grand!

felddy committed:
Apply correct typographic designation to character replacement comment.

jsf9k commented on .github/workflows/build.yml:
```diff
@@ -46,40 +69,179 @@ jobs:
+    #
+    #     Builds targeting the default branch will be tagged with `:edge`.
+    #
+    #     Builds from other branches will be tagged with the branch name. Solidi
```
I'm going to insist that we change `Solidi` to `Solidi ("slash")`, and I suggest that we provide a link for solidus here.  This is some pretty arcane terminology, which I love, but we need to offer the reader:
* The vernacular word, so he or she is not lost
* A link to wikipedia or whatever for his or her edification

Perhaps our readers will one day become so numerous that "solidus" becomes a commonly known word (like "octothorpe") and we can remove the parenthesized vernacular.

jsf9k replied:
**Solidus**, the guitarist from GNR
![image](https://user-images.githubusercontent.com/5521725/98273602-737faa00-1f60-11eb-98e4-0aa2e220a582.png)

jsf9k replied:
[This](https://en.wikipedia.org/wiki/Slash_(punctuation)#Encoding) looks to be an excellent link to use.

jsf9k reviewed (changes requested):
I am requesting one minor change, for clarity.

felddy committed:
Modernize Docker shields in README to point to working links.

felddy commented on .github/workflows/build.yml:
```diff
@@ -46,40 +69,179 @@ jobs:
+    #
+    #     Builds targeting the default branch will be tagged with `:edge`.
+    #
+    #     Builds from other branches will be tagged with the branch name. Solidi
+    #     `/` in branch names are replaced with hyphen-minuses `-` in the Docker
+    #     tag.
```
@jsf9k Glad you're enjoying the Solidus! 😝

Not sure if you saw the ```(`\`)``` on the next comment line...  Are you suggesting:
```suggestion
    #     Builds from other branches will be tagged with the branch name. Solidi
    #     ("slahes") in branch names are replaced with hyphen-minuses `-` in the Docker
    #     tag.  See: https://www.compart.com/en/unicode/U+002F
```
-or-
```suggestion
    #     Builds from other branches will be tagged with the branch name. Solidi
    #     ("slahes") `/` in branch names are replaced with hyphen-minuses `-` in the Docker
    #     tag.  See: https://www.compart.com/en/unicode/U+002F
```

jsf9k replied:
I did miss the solidus that followed.  How about this:
```suggestion
    #     Builds from other branches will be tagged with the branch name. Solidi
    #     (`/` characters - commonly known as slashes) in branch names are replaced
    #     with hyphen-minuses (`-` characters) in the Docker tag.  For more information
    #     about the solidus see these links:
    #       * https://www.compart.com/en/unicode/U+002F
    #       * https://en.wikipedia.org/wiki/Slash_(punctuation)#Encoding
```

felddy replied:
Love it.  Someone is going to wonder what makes this workflow tick, and they're gonna get a typography schooling for free!

jsf9k replied:
There is a thang in thermodynamics called [_fugacity_](https://en.wikipedia.org/wiki/Fugacity).  It is also an English word that means "fleeting".  What I _most_ remember from learning about fugacity in grad school was the footnote at the bottom of the page where it was introduced in our textbook.  It was an excerpt from a quote by Samuel Johnson - "...the fugacity of pleasure, the fragility of beauty...".

Sometimes these asides really tie the room together.

jsf9k replied:
> Add some additional comments in case anyone is confused about the nature of the solidus and its place within our tagging system.

What a great comment!

This whole exchange reminds me of [the Charles Portis book "Masters of Atlantis"](https://en.wikipedia.org/wiki/Masters_of_Atlantis) and Lamar Jimmerson and Sydney Hen's very serious study of the gnomon.  The solidus is the dev team's gnomon!

felddy committed:
Elaborate on the Solidus.

Add some additional comments in case anyone is confused about the nature of the solidus and its place within our tagging system.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k reviewed (approved):
Thank you for elaborating upon the solidus.  We must educate the masses!
